Isaac/signup flow

ittullos opened the pull request:
This PR adds the ability to create a new user with the email address from the ID token. I renamed some variables to help distinguish between the access token and id token. I'm not 100% sure but it seems like we don't even need a /signup route. I've changed my signup code on the front end to pass both tokens and it seems to be working. The last commit updates the user's email if it doesn't not match the email in the token. This should work to update the user after an email change. I feel like this commit should have some additional tests to go with it but I was a little confused when working with the authentication_spec. I'm also curious if you think there are any situations where the authentication code could run into errors or if there are any vulnerabilities with the way things are set up. Thanks!!

ittullos committed:
Makes first and last name optional for user

This changes the User model to allow for users to not have a first
or last name. This makes it more simple for new user's who are signing
up. We can add an info section to the settings screen where they can
add their first and last name if they want to. We could also add a pop
up when they first sign up.

ittullos committed:
Adds ID token functionality to authentication middleware

Makes it so that the authentication middleware can accept an ID token
from the client and use it to obtain the user's email address. The
email address is then used to create a new user in the database if one
does not already exist.

ittullos committed:
Add email change functionality

This commit adds the ability for users to change their email. With
every request, the ID token is checked to see if the email has been
changed. If it has, the user record in the database will be updated.

zbarnard1 commented on spec/middleware/authentication_spec.rb:
```diff
@@ -45,6 +45,30 @@
+
+          response = subject.call(env)
+          expect(response).to eq([200, {}, 'success'])
+          expect(User.last.email).to include("isaac.tullos")
```
Shouldn't this be Gandalf instead of Isaac? Gandalf is the email in the Id Token, so I'm assuming that should get set on the user and not the original email.

ittullos replied:
Yeah this part was really confusing to me. The tests failed because the actual email was the isaac.tullos one from the factory. Not sure why it would be that instead of gandalf.

zbarnard1 commented on app/middleware/authentication.rb:
```diff
@@ -7,11 +7,20 @@ def initialize(app)
+
+      if id_token
+        id_token_json = JSON.parse(id_token)
+        email = id_token_json.dig('payload', 'email')
```
Is this `payload` key correct? I thought we had a flatter structure for the token. I don't see a `payload` in the id_token_double in the test so this is confusing.

ittullos replied:
![Screenshot 2025-02-04 at 12 02 41 AM (2)](https://github.com/user-attachments/assets/0c753816-565c-44ba-bbd1-e92b528a29b1)

ittullos replied:
This is what the structure of the real ID token that I'm getting from my front end looks like. Should I make it so my test token matches this structure?

ittullos replied:
Here's what the real ID token that I'm getting from my front end looks like. Should I be making it so that my id_token_double has this same structure? Looks like the payload that contains the email is not inside the JWT

![Screenshot 2025-02-04 at 12 08 29 AM (2)](https://github.com/user-attachments/assets/5eba54ba-acbc-48d7-ac66-31b2241adc57)

zbarnard1 replied:
that `jwtToken` key isn't standard, this looks like the data that's in the javascript object that wraps the token. You're going to send the value of `jwtToken` and it's going to get decoded by the server into the `payload`, so we won't have that `payload` key.

ittullos replied:
Ahhh okay, just decoded the JWT from the id token and the email is contained inside! For some reason I thought it was only inside that payload object! This def has me on the right track now. Thanks!! 🙌

zbarnard1 commented on app/middleware/authentication.rb:
```diff
@@ -23,6 +23,13 @@ def call(env)
       end
 
       if user
+        if email
```
This is a good spot to extract a well-named method to describe the operation that's happening. We could pass some of these values in as arguments and return if any are nil instead of doing a nested conditional. You can also do something like `if user.present? && email.present?` instead of nesting.

zbarnard1 reviewed (changes requested):
I feel like we've got a little work to do on the authentication spec. I don't think this is working as intended right now because it seems like the email should be changing but it's not.

ittullos committed:
Adds id_token to authentication spec helpers

This commit adds the id_token to the authentication spec helpers
so that the id_token can be required in the authentication middleware
without causing the tests to fail. The tests have also been changed to
test a new user as well as a user who's email address has changed. The
authentication code has also changed to take in only the JWT and not
the entire idToken object.

There are currently 2 failing tests in the route_spec and I can't seem
to figure out why. It doesn't look like the headers are coming through
with the request at all. I will definitely need to fix this before
merging.

ittullos commented:
So I changed the code in the authentication middleware to take in only the JWT instead of the entire idToken object. I kept running into issues with having to guard the id_token, since adding it into the authentication middleware made several of the tests fail. I decided to go ahead and add the id_token to the authentication spec helpers and change all the test accordingly. I had 33 failing tests and I've worked it down to 2. They are both in the route_spec and I don't see any headers coming through with the request at all. I'll leave a comment where the failing tests are. Thanks!

ittullos commented on spec/requests/route_spec.rb:
```diff
@@ -18,7 +18,7 @@
     it 'creates a Route' do
       expect do
-        post '/user/routes', {}, user_token_header(user, { 'sub' => user.sub })
+        post '/user/routes', {}, headers
```
Failing here: `expected `Route.count` to have changed by 1, but was changed by 0`

ittullos commented on spec/requests/route_spec.rb:
```diff
@@ -28,7 +28,7 @@
     it 'creates a Route with a nil commitment' do
       allow_any_instance_of(User).to receive(:current_commitment).and_return(nil)
-      post '/user/routes', {}, user_token_header(user, { 'sub' => user.sub })
+      post '/user/routes', {}, headers
```
Failing here: `expected: 200 got: 401`

zbarnard1 commented on app/middleware/authentication.rb:
```diff
@@ -7,11 +7,20 @@ def initialize(app)
 
-      decoded_token = JsonWebToken.new(token).verify!
-      user = User.find_or_create_by(sub: decoded_token.fetch('sub'))
+      user = User.find_or_create_by(sub: decoded_access_token.fetch('sub')) do |u|
```
⛏️ We can use `find_or_create_by!` and drop the `save!`

zbarnard1 commented on app/middleware/authentication.rb:
```diff
@@ -11,25 +11,18 @@ def call(env)
-            user.save!
-          end
-        end
+      update_user_email_if_needed(user, email)
```
We could move this down into the conditional below and not have to check the user twice, see my other comment :)

zbarnard1 commented on app/middleware/authentication.rb:
```diff
@@ -41,6 +34,15 @@ def call(env)
     private
 
+    def update_user_email_if_needed(user, email)
+      return unless user.present? && email.present?
```
If we move this method down a couple lines so it's inside the conditional on line 25, we can drop the `user.present?`.
Also, we know that the email exists by this point, or there was an "email" key in the Id token, anyway. If this value is nil or an empty string, then we won't be able to save this record. It will already have raised a validation error if the user is new, and later we'll be using `save!` so we won't be setting the email to an invalid value if we're updating an existing user.

zbarnard1 commented on app/middleware/authentication.rb:
```diff
@@ -41,6 +34,15 @@ def call(env)
+      return unless user.present? && email.present?
+
+      if user.email != email
+        user.email = email
```
ActiveRecord has a cool way of doing this which I just wanted to show you, you don't necessarily have to implement it this way.
```ruby
 user.email = email
 user.save! if user.changed?
```

ittullos replied:
Very cool!!

zbarnard1 reviewed (approved):
Works for me! I'm approving this with the assumption you'll rebase my little fixup and you can do any style tweaks you want.
Nice work!
